packages/sanity/src/core/form/studio/tree-editing/components/TreeEditingDialog.tsx

```diff
@@ -0,0 +1,254 @@
+    [handleBuildTreeEditingState],
+  )
+
+  const onClose = useCallback(() => {
+    // Cancel any debounced state building when closing the dialog.
+    debouncedBuildTreeEditingState.cancel()
+
+    // Reset the tree state when closing the dialog.
+    setTreeState(EMPTY_TREE_STATE)
+
+    // Reset the `openPath`
+    onPathOpen(EMPTY_ARRAY)
+
+    // Reset the stored value and openPath to undefined.
+    // This is important since the next time the dialog is opened,
+    // we want to build the tree editing state from scratch and
+    // don't prevent the state from being built by comparing the
+    // previous stored values.
+    valueRef.current = undefined
+    openPathRef.current = undefined
+  }, [debouncedBuildTreeEditingState, onPathOpen])
+
+  const open = useMemo(
+    () => shouldArrayDialogOpen(schemaType, treeState.relativePath),
+    [treeState.relativePath, schemaType],
+  )
+
+  const onHandlePathSelect = useCallback(
+    (path: Path) => {
+      // Cancel any debounced state building when navigating.
+      debouncedBuildTreeEditingState.cancel()
+
+      onPathOpen(path)
+
+      // If the path is not an array item path, it means that the field is
+      // present in the form. In that case, we want to focus the field
+      // in the form when it is selected in order to scroll it into view.
+      if (!isArrayItemPath(path)) {
+        onPathFocus(path)
+      }
+    },
+    [debouncedBuildTreeEditingState, onPathFocus, onPathOpen],
+  )
+
+  useEffect(() => {
+    // Don't proceed with building the tree editing state if the dialog
+    // should not be open.
+    if (!shouldArrayDialogOpen(schemaType, openPath)) return
+
+    const valueChanged = !isEqual(value, valueRef.current)
+    const openPathChanged = !isEqual(openPath, openPathRef.current)
+    const isInitialRender = valueRef.current === undefined && openPathRef.current === undefined
+
+    // If the value has not changed but the openPath has changed, or
+    // if it is the initial render, build the tree editing state
+    // without debouncing. We do this to make sure that the UI is
+    // updated immediately when the openPath changes.
+    // We only want to debounce the state building when the value changes
+    // as that might happen frequently when the user is editing the document.
+    if (isInitialRender || openPathChanged) {
+      handleBuildTreeEditingState({
+        schemaType,
+        documentValue: value,
+        openPath,
+      })
+
+      openPathRef.current = openPath
+
+      return
+    }
+
+    // Don't proceed with building the tree editing state if the
+    // openPath and value has not changed.
+    if (!valueChanged && !openPathChanged) return
+
+    // Store the openPath and value to be able to compare them
+    // with the next openPath and value.
+    valueRef.current = value
+    openPathRef.current = openPath
+
+    debouncedBuildTreeEditingState({
+      schemaType,
+      documentValue: value,
+      openPath,
+    })
```

Comment:
❓ should there also be a debounce cancel like in the close handler? 

```tsx
return () => {
 // Cancel any debounced state building when navigating.
      debouncedBuildTreeEditingState.cancel()
}
```

Reply:
Yeah this definitely sounds like a good idea

Reply:
Though I see tht we do have a debounce already on the onClose :) 
https://github.com/sanity-io/sanity/blob/edx-1295/packages/sanity/src/core/form/studio/tree-editing/components/TreeEditingDialog.tsx#L115